eventsources/sources/bitbucketserver/start.go

```diff
@@ -0,0 +1,322 @@
+			return errors.Errorf("failed to delete webhook. err: %+v", err)
+		}
+
+		logger.Info("bitbucket server webhook deleted")
+	}
+	return nil
+}
+
+func (router *Router) CreateBitbucketWebhook(ctx context.Context, bitbucketConfig *bitbucketv1.Configuration) error {
+	bitbucketserverEventSource := router.bitbucketserverEventSource
+	route := router.route
+
+	logger := route.Logger.With(
+		logging.LabelEndpoint, route.Context.Endpoint,
+		logging.LabelPort, route.Context.Port,
+		logging.LabelHTTPMethod, route.Context.Method,
+		"project-key", bitbucketserverEventSource.ProjectKey,
+		"repository-slug", bitbucketserverEventSource.RepositorySlug,
+		"base-url", bitbucketserverEventSource.BitbucketServerBaseURL,
+	)
+
+	router.bitbucketClient = bitbucketv1.NewAPIClient(ctx, bitbucketConfig)
+
+	formattedURL := common.FormattedURL(bitbucketserverEventSource.Webhook.URL, bitbucketserverEventSource.Webhook.Endpoint)
+
+	apiResponse, err := router.bitbucketClient.DefaultApi.FindWebhooks(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, nil)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to list existing hooks to check for duplicates for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	hooks, err := bitbucketv1.GetWebhooksResponse(apiResponse)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert the list of webhooks for repository %s/%s", router.bitbucketserverEventSource.ProjectKey, router.bitbucketserverEventSource.RepositorySlug)
+	}
+
+	var existingHook bitbucketv1.Webhook
+	isAlreadyExists := false
+
+	for _, hook := range hooks {
+		if hook.Url == formattedURL {
+			existingHook = hook
+			isAlreadyExists = true
+		}
+	}
+
+	logger.Info("retrieving the webhook secret...")
+	webhookSecret, err := router.getWebhookSecret(bitbucketserverEventSource.WebhookSecret)
+	if err != nil {
+		return errors.Errorf("failed to get bitbucketserver webhook secret. err: %+v", err)
+	}
+
+	newHook := bitbucketv1.Webhook{
+		Name:          "Argo Events",
+		Url:           formattedURL,
+		Active:        true,
+		Events:        bitbucketserverEventSource.Events,
+		Configuration: bitbucketv1.WebhookConfiguration{Secret: webhookSecret.secret},
+	}
+
+	localVarPostBody, err := json.Marshal(newHook)
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal new webhook to JSON")
+	}
+
+	if !isAlreadyExists {
+		apiResponse, err = router.bitbucketClient.DefaultApi.CreateWebhook(bitbucketserverEventSource.ProjectKey, bitbucketserverEventSource.RepositorySlug, localVarPostBody, []string{"application/json"})
+
+		if err != nil {
+			return errors.Errorf("failed to add webhook. err: %+v", err)
+		}
+
+	} else {
+		logger.Info("webhook already exists, updating it...")
```

Comment:
I'm not familiar with bitbucket, but have a feeling that calling the update API once (or multiple times if replicas > 1) a minute sounds not good.